server/sonar-server/src/main/java/org/sonar/server/issue/ws/ChangelogAction.java

```diff
@@ -0,0 +1,157 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.issue.ws;
+
+import com.google.common.io.Resources;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.core.issue.FieldDiffs;
+import org.sonar.core.util.stream.Collectors;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.issue.IssueDto;
+import org.sonar.db.user.UserDto;
+import org.sonar.server.issue.IssueFinder;
+import org.sonarqube.ws.Issues.ChangelogWsResponse;
+import org.sonarqube.ws.Issues.ChangelogWsResponse.Changelog;
+
+import static com.google.common.base.Strings.emptyToNull;
+import static org.sonar.api.utils.DateUtils.formatDateTime;
+import static org.sonar.core.util.Protobuf.setNullable;
+import static org.sonar.core.util.Uuids.UUID_EXAMPLE_01;
+import static org.sonar.server.issue.IssueFieldsSetter.TECHNICAL_DEBT;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.issue.IssuesWsParameters.ACTION_CHANGELOG;
+import static org.sonarqube.ws.client.issue.IssuesWsParameters.PARAM_ISSUE;
+
+public class ChangelogAction implements IssuesWsAction {
+
+  private static final String EFFORT_CHANGELOG_KEY = "effort";
+
+  private final DbClient dbClient;
+  private final IssueFinder issueFinder;
+
+  public ChangelogAction(DbClient dbClient, IssueFinder issueFinder) {
+    this.dbClient = dbClient;
+    this.issueFinder = issueFinder;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_CHANGELOG)
+      .setDescription("Display changelog of an issue<br/>." +
+        "Require the 'Browse' permission on the project of the specified issue<br/>." +
+        "Since 6.3, changes on effort are returning raw value in minutes, it doesn't return anymore the duration.")
+      .setSince("4.1")
+      .setHandler(this)
+      .setResponseExample(Resources.getResource(IssuesWs.class, "example-changelog.json"));
+    action.createParam(PARAM_ISSUE)
+      .setDescription("Key of the issue")
+      .setRequired(true)
+      .setExampleValue(UUID_EXAMPLE_01);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Stream.of(request)
+        .map(searchChangelogs(dbSession))
+        .map(buildResponse())
+        // TODO use Collectors.toOneElement to improve readability
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)